lib/rules/generator-star-spacing.js

```diff
@@ -126,17 +171,23 @@ module.exports = {
             }
 
             const starToken = getStarToken(node);
-
-            // Only check before when preceded by `function`|`static` keyword
             const prevToken = sourceCode.getTokenBefore(starToken);
+            const nextToken = sourceCode.getTokenAfter(starToken);
 
-            if (prevToken.value === "function" || prevToken.value === "static") {
-                checkSpacing("before", prevToken, starToken);
+            let kind = "named";
+
+            if (node.parent.type === "MethodDefinition" || (node.parent.type === "Property" && node.parent.method)) {
+                kind = "method";
+            } else if (!node.id) {
+                kind = "anonymous";
             }
 
-            const nextToken = sourceCode.getTokenAfter(starToken);
+            // Only check before when preceded by `function`|`static` keyword
+            if (kind !== "method" || prevToken.value === "static") {
```

Comment:
I'm wondering if this check could be expressed in a better way. Conceptually, it seems like the case we care about isn't whether the value of the previous token happens to be `static` -- it's that we don't want to enforce spacing before the star when it's the first token of the given property.

```js
class Foo {
  *bar() {} // '*' is the first token of the MethodDefinition
}

class Foo {
  static bar() {} // '*' is not the first token of the MethodDefinition (it's preceded by `static`)
}

// (experimental syntax)
class Foo {
  async *bar(){} // '*' is not the first token of the MethodDefinition (it's preceded by `async`)
}
```

So it might be better to express this statement as

```js
if (!(kind === "method" && starToken === sourceCode.getFirstToken(node.parent))) {
    checkSpacing(kind, "before", prevToken, starToken);
}
```

This matters because although we don't explicitly support specific custom/experimental syntax in core rules, the rule will probably be run with various custom parsers, so it's useful to make the rule robust when dealing with unrecognized syntax in general. To that end, it helps to make rule's conditionals semantically correct so that they behave reasonably when they encounter syntax they don't know how to deal with.